### PR TITLE
Add `property.Glob` & implement `encoding.(Unm|M)arshalText` for `property.Path`

### PR DESCRIPTION
**Pull request by iwahbe**
Adds a new `property.Glob` to express a property path that can contain globs. This PR adds to & from text conversions for `property.Glob` & `property.Path`.

This is in preparation for porting the implementation of property path parsing to `property` to tighten up distinguishing `[*]` from `["*"]`.

`property.Glob` is, and `property.Path` has been converted to, a `string` backed binary format in memory. This gives us type based immutability and lets us put a `property.Path` or `property.Glob` into a native `map` type as a key.

### CODE REVIEW

File: sdk/go/property/fmt_test.go [hunk 1]
```diff
@@ -117,3 +117,49 @@ func TestGoStringMap(t *testing.T) {
 		assert.Equal(t, expt, NewMap(map[string]Value{"k1": New(true), "k2": New(false)}).GoString())
 	})
 }
+
```
Comment (Frassle) [inline sdk/go/property/fmt_test.go:120]: String() tests?

File: sdk/go/property/path_test.go [hunk 2]
```diff
@@ -116,46 +116,31 @@ func TestGet(t *testing.T) {
...
-			from: property.New([]property.Value{
-				property.New("0"),
-			}),
-			failure: &pathFailure{
-				found: property.New([]property.Value{
-					property.New("0"),
-				}),
-				msg: "index -1 out of bounds of an array of length 1",
-			},
-		},
```
Comment (Frassle) [inline sdk/go/property/path_test.go:133]: Is this impossible to construct now?

File: sdk/go/property/path_repr.go [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+func (p pathRepr) appendKey(s string) pathRepr {
+	contract.Assertf(len(s) < 1<<14, "string exceeded max length")
+	var hdr [2]byte
+	binary.BigEndian.PutUint16(hdr[:], uint16(2<<14|len(s))) //nolint:gosec // checked above
+	return pathRepr{p.string + string(hdr[:]) + s}
+}
+
+func (p pathRepr) appendIndex(i int) pathRepr {
+	contract.Assertf(i < 1<<14, "index exceeded max length")
+	contract.Assertf(i >= 0, "index must be non-negative")
```
Comment (Frassle) [inline sdk/go/property/path_repr.go]: "x[-1]" isn't a valid path for parsing right, so while we could make an IndexSegment in code we'd have to manually put in a negative right? So I think that's fine.
14 bits is a bit more worrying, I think the path parser checks for "fits in `int`" (which I think is 32bit... but not sure Go actually defines that strictly). This is extremely unlikely but I'd rather avoid a panic.

File: sdk/go/property/fmt_test.go (hunk 1 above)
Comment (iwahbe) [reply sdk/go/property/fmt_test.go:120]: They purposely don't implement `String`, since its not obvious what `Path{}` or `Glob{}` should return. `(Path{}).MarshalText()` will error since it's not valid to marshal. I'm undecided what `(Path{}).String()` should return: `[empty]` or `""`?

File: sdk/go/property/path_test.go (hunk 2 above)
Comment (iwahbe) [reply sdk/go/property/path_test.go:133]: Yes. I'll add a test showing that unmarshaling a negative or over-large number will error.

File: sdk/go/property/fmt_test.go (hunk 1 above)
Comment (Frassle) [reply sdk/go/property/fmt_test.go:120]: ack that makes sense

File: sdk/go/property/path_repr.go (hunk 3 above)
Comment (iwahbe) [reply sdk/go/property/path_repr.go]: The HCL parser actually parses into `big.Float`. I'll adjust the system to handle `u32`s. I think that should be sufficient, since that's the max size a go array can handle.

Comment (Frassle) [reply sdk/go/property/path_repr.go]: I was looking at our `parsePropertyPath` which does `index, err := strconv.ParseInt(segment, 10, 0)`

So that's the set of integers that users could potentially have passed to the engine via ignoreChanges, etc, that we shouldn't start panic'ing on.

Comment (iwahbe) [reply sdk/go/property/path_repr.go]: A bit size of `0` (parsing into `int`) means we should be safe to parse into an `uint32` since we know that we only need the positive half of `int` (`int64` on modern systems).

Comment (Frassle) [reply sdk/go/property/path_repr.go]: uint32 < int63 😕

Comment (iwahbe) [reply sdk/go/property/path_repr.go]: Wow, that was dumb, sorry. Switching to a 64 bit number now.

File: sdk/go/property/fmt.go [hunk 4]
```diff
@@ -101,3 +104,44 @@ func (a Map) GoString() string {
 func (null) GoString() string { return "property.Null" }
 
 func (computed) GoString() string { return "property.Computed" }
+
+func (g Path) String() string { return g.AsGlob().String() }
```
Comment (Frassle) [inline sdk/go/property/fmt.go]: I thought we weren't adding `String`? What changed your mind?

File: sdk/go/property/glob.go [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+		}
+		if len(runes) < i || runes[i] != ']' {
+			return nil, nil, fmt.Errorf("unclosed index [%s", string(runes[:i]))
+		}
+		n, err := strconv.ParseUint(string(runes[0:i]), 10, 64)
+		if err != nil {
+			return nil, nil, err
+		}
+		if n > math.MaxInt64 {
+			return nil, nil, errors.New("indexes cannot exceed 9223372036854775807")
```
Comment (Frassle) [inline sdk/go/property/glob.go]:
```suggestion
			return nil, nil, errors.New("indexes cannot exceed %d", int64(math.MaxInt64))
```
?

File: sdk/go/property/path.go [hunk 6]
```diff
@@ -158,15 +199,21 @@ func (k KeySegment) apply(v Value) (Value, PathApplyFailure) {
...
+type IndexSegment struct {
+	// i must be constrained into a uint63, since it must be cleanly castable to an [int] ([int64]).
+	i uint64
+}
+
+func (k IndexSegment) Index() int { return int(k.i) } //nolint:gosec // will always fit
 
 func (k IndexSegment) apply(v Value) (Value, PathApplyFailure) {
 	if v.IsArray() {
+		i := int(k.i) //nolint:gosec // will always fit
```
Comment (Frassle) [inline sdk/go/property/path.go:211]: Not sure that's totally to spec? `int` is machine width size, I don't think we ever build for 32 bits though

Comment (iwahbe) [reply sdk/go/property/path.go:211]: It's not listed explicitly in https://www.pulumi.com/docs/get-started/download-install/#minimum-system-requirements, but we never build for a 32 bit system.

File: sdk/go/property/fmt.go (hunk 4 above)
Comment (iwahbe) [reply sdk/go/property/fmt.go]: Being sick I guess. I stand by my earlier logic and have removed the `String` functions.
